_posts/2017-03-28-table-sql-api-update.md

```diff
@@ -0,0 +1,195 @@
+
+```scala
+class Customer {
+  var id: Int = _
+  var name: String = _
+  var update: Long = _
+  var prefs: java.util.Properties = _
+}
+```
+We can use the following table program to convert the CSV file into Customer objects. Flink takes care of creating objects and mapping fields for us.
+
+```scala
+val ds = tEnv
+  .scan("customers")
+  .select('id, 'name, 'last_update as 'update, parseProperties('prefs) as 'prefs)
+  .toDataStream[Customer]
+```
+
+Even though Flink’s Table API is shipped with a large set of built-in functions, it is often necessary to define custom user-defined scalar functions in this example to parse the preferences into a Java Properties object. The following code snippet shows how easily we can implement a scalar function.
+
+```scala
+object parseProperties extends ScalarFunction {
+  def eval(str: String): Properties = {
+    val props = new Properties()
+    str
+      .split(",")
+      .map(\_.split("="))
+      .foreach(split => props.setProperty(split(0), split(1)))
+    props
+  }
+}
+```
+
+Scalar functions can be used to deserialize, extract, or convert values (and more). By overwriting the `open()` method we can even have access to runtime information such as distributed cached files or metrics. Even the `open()` method is only called once during the runtime’s [task lifecycle](https://ci.apache.org/projects/flink/flink-docs-release-1.3/internals/task_lifecycle.html).
+
+## Unified Windowing for Static and Streaming Data
+
+Another very common task, especially when working with continuous data, is the definition of windows to split a stream into pieces of finite size, over which we can apply computations. At the moment, the Table API supports three types of windows: sliding windows, tumbling windows, and session windows (for general definitions of the different types of windows, we recommend [Flink’s documentation](https://ci.apache.org/projects/flink/flink-docs-release-1.2/dev/windows.html)). All three window types work on [event or processing time](https://ci.apache.org/projects/flink/flink-docs-release-1.2/dev/event_time.html). Session windows can be defined over time intervals, sliding and tumbling windows can be defined over time intervals or a number of rows.
+
+Let’s assume that our customer data from the example above is an event stream of updates generated whenever the customer updated his or her preferences. We assume that events come from a TableSource that has assigned timestamps and watermarks. The definition of a window happens again in a LINQ-style fashion. The following example could be used to count the updates to the preferences during one day.
+
+```scala
+table
+  .window(Tumble over 1.day on 'rowtime as 'w)
+  .groupBy('id, 'w)
+  .select('id, 'w.start as 'from, 'w.end as 'to, 'prefs.count as 'updates)
+```
+
```

Comment:
The point that timestamps & watermarks have to be assigned is mentioned several times.